metrics/sinks/stackdriver/stackdriver.go

```diff
@@ -358,27 +351,74 @@ func CreateStackdriverSink(uri *url.URL) (core.DataSink, error) {
 		}
 	}
 
-	if err := gce_util.EnsureOnGCE(); err != nil {
-		return nil, err
+	var projectId, heapsterZone string
+	// Cluster name and location are required when useNewResourceModel is true.
+	var clusterName, clusterLocation string
+
+	if len(opts["cluster_name"]) >= 1 {
+		clusterName = opts["cluster_name"][0]
 	}
 
-	// Detect project ID
-	projectId, err := gce.ProjectID()
-	if err != nil {
-		return nil, err
+	if len(opts["cluster_location"]) >= 1 {
+		clusterLocation = opts["cluster_location"][0]
 	}
 
-	// Detect zone for old resource model
-	heapsterZone, err := gce.Zone()
-	if err != nil {
-		return nil, err
+	if gce.OnGCE() {
+		// Detect project ID
+		projectId, err = gce.ProjectID()
+		if err != nil {
+			return nil, err
+		}
+
+		// Detect zone for old resource model
+		heapsterZone, err = gce.Zone()
+		if err != nil {
+			return nil, err
+		}
+
+		if useNewResourceModel {
+			if clusterName == "" {
+				glog.Info("An empty cluster name has been provided, checking the GCE Metadata Server to try to auto-detect.")
+
+				clusterName_, err := gce.InstanceAttributeValue("cluster-name")
+				if err == nil {
+					glog.Infof("Discovered '%s' as the cluster name from the GCE Metadata Server.", clusterName_)
+					clusterName = clusterName_
+				} else {
+					glog.Warningf("Cluster name could not be discovered using the GCE Metadata Server: %s", err)
+				}
+			}
+
+			if clusterLocation == "" {
+				glog.Info("An empty cluster location has been provided, checking the GCE Metadata Server to try to auto-detect.")
+
+				clusterLocation_, err := gce.InstanceAttributeValue("cluster-location")
+				if err == nil {
+					glog.Infof("Discovered '%s' as the cluster location from the GCE Metadata Server.", clusterLocation_)
+					clusterLocation = clusterLocation_
+				} else {
+					glog.Warningf("Cluster location could not be discovered using the GCE Metadata Server: %s", err)
+				}
+			}
+		}
+	} else {
+		projectId, err = gce_util.GetProjectId()
```

Comment:
Should we have a comment here as well, e.g., `// Get project ID from the environment`?

Reply:
Done